backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/tournament/TournamentService.java

```diff
@@ -0,0 +1,343 @@
+            throw new TutorException(TOURNAMENT_CANCELED, tournament.getId());
+        }
+
+        if (tournament.getParticipants().contains(user)) {
+            throw new TutorException(DUPLICATE_TOURNAMENT_PARTICIPANT, user.getUsername());
+        }
+
+        if (!user.getCourseExecutions().contains(tournament.getCourseExecution())) {
+            throw new TutorException(STUDENT_NO_COURSE_EXECUTION, user.getId());
+        }
+
+        if (tournament.isPrivateTournament() && !password.equals(tournament.getPassword())) {
+            throw new TutorException(WRONG_TOURNAMENT_PASSWORD, tournament.getId());
+        }
+
+        tournament.addParticipant(user);
+
+        if (!tournament.hasQuiz()) {
+            StatementCreationDto quizForm = new StatementCreationDto();
+            quizForm.setNumberOfQuestions(tournament.getNumberOfQuestions());
+            quizForm.setAssessment(tournament.getAssessment().getId());
+
+            StatementQuizDto statementQuizDto = statementService.generateStudentQuiz(tournament.getCreator().getId(),
+                    tournament.getCourseExecution().getId(), quizForm);
+
+            Quiz quiz = quizRepository.findById(statementQuizDto.getId())
+                    .orElseThrow(() -> new TutorException(QUIZ_NOT_FOUND, statementQuizDto.getId()));
+
+            if (DateHandler.now().isBefore(tournament.getStartTime())) {
+                quiz.setAvailableDate(tournament.getStartTime());
+            }
+            quiz.setConclusionDate(tournament.getEndTime());
+            quiz.setTitle("Tournament " + tournament.getId() + " Quiz");
+            quiz.setType(Quiz.QuizType.TOURNAMENT.toString());
+
+            tournament.setQuizId(statementQuizDto.getId());
+        }
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public StatementQuizDto solveQuiz(Integer userId, TournamentDto tournamentDto) {
+        User user = checkUser(userId);
+        Tournament tournament = checkTournament(tournamentDto.getId());
+
+        if (!tournament.getParticipants().contains(user)) {
+            throw new TutorException(USER_NOT_JOINED, userId);
+        }
+
+        if (!tournament.hasQuiz()) {
+            throw new TutorException(TOURNAMENT_NO_QUIZ, tournamentDto.getId());
+        }
+
+        return statementService.startQuiz(userId, tournament.getQuizId());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public void leaveTournament(Integer userId, TournamentDto tournamentDto) {
+        User user = checkUser(userId);
+        Tournament tournament = checkTournament(tournamentDto.getId());
+
+        if (!tournament.getParticipants().contains(user)) {
+            throw new TutorException(USER_NOT_JOINED, user.getUsername());
+        }
+
```

Comment:
It is not necessary to verity that it did not answer any question, or that the quiz did not start``